### ISSUE TITLE
Issue with the Calibre plugin on FFnet

### ISSUE THREAD

**Issue by Twilight666** [JimmXinu/FanFicFare#163]
On my computer I use both the CLI and the calibre versions.

The calibre I use as a library of my favorite fanfics and the CLI to download new stories. If I like the story enough I add it to my calibre version

I noticed that some times when a story is updated and I try to update the story through calibre, if I pick update (or overwrite) if newer then calibre tells me that it is up-to-date when it clearly is not.
If I force it to update/overwrite then those times it tends to not download the latest chapter.
I should note that I have noticed this on fanfiction.net since almost all my stories in calibre are from there.

The above is not a common issue and it goes away if I wait like 10 minutes, so it is probably a very difficult to verify and fix bug since it is something temporary.
On top of that there is a very good possibility that it is an issue on FFnet. I don't know how you download stories from ffnet but I guess you create some kind of index and download the individual chapters so it is possible that it just takes time for FFnet to make the index up-to-date and while the new chapter is up on the site the index does not point to it and gets skipped.

However I decided to check something. A couple of the stories on Calibre updated today and the overwrite if newer option was telling me bith times that the story had not updated. So I tried to force them to download from calibre while at the same time use the CLI plugin and compare the results. I did that since I had not noticed something like that when I download stories through the CLI.

I used 2 ways to download the stories through calibre.
The first time I got an alert that a story updated I went to calibre and tried to update it. It told me that it was up to date, so I checked the option to overwrite always. At the same time I downloaded the story through the CLI. The calibre version did not get the latest chapter while the CLI was up-to-date.
The second time it happened I decided to delete the story and redownload it using the exact same url as the CLI since that pointed to the latest chapter. Again the calibre version was missing the latest chapter while the CLI was up-to-date.
I should note that the CLI version was started _before_ the calibre version to avoid the possibility that the delay might cause a difference.
Also I have not noticed a simiar issue on the CLI but I should not that the url I use for CLI points to the latest chapter (I don't know if that affects anything) which is why I chose to use that for the second story.

I don't know if the above if fixable since it might be a FFnet issue that you can't do anything about and it is difficult to check and fix in any case but I figured since it is a somewhat reoccuring issue, I should probably mention it since it only seems to happen on calibre and not on CLI

**Comment by JimmXinu** [comment]
It sounds like you're seeing a known issue: The ffnet servers don't all update at the same time.  It's an issue I've seen be more and less common over the years.

There's already an optional feature implemented that can help.  It's off by default, but you can turn it on in your personal.ini:

```
[www.fanfiction.net]
## ffnet used to have a tendency to send out update notices in email
## before all their servers were showing the update on the first
## chapter.  It generates another server request and doesn't seem to
## be needed lately, so now default it to off.
check_next_chapter:true
```

FanFicFare will look at the chapter list it gets from the first chapter and looks for one chapter past the end.  I can't promise that will work all the time, but it should help.

**Comment by Twilight666** [comment]
I am not sure if I should make a new issue but since this is still about the same thing:

Ok I have been using the fix and it seems to be working properly.
However that only applies if I choose to force the calibre plugin to download the epub again. That way it does the check you say _during_ the download

My default option is to download if newer exists though and in those cases during the check it does not recognise that there is a new chapter and does not start the download.
In fact I tried the following:
I force calibre to download the story.
I wait for the download to end and for calibre to show me the download complete message.
I try to update the story at which point I get the "none of the stories need to be downloaded message"
I then press yes on the download complete message from the previous attempt and the new chapter is there!!

Is there a way to do something about that?

**Comment by JimmXinu** [comment]
If you didn't click 'yes' to 'Proceed with updating your library' after the first download finished, then it (surprise!) hasn't updated your library yet. :) That's why you only see the new chapter afterwards.

Why does 'overwrite always' get the new chapter and 'overwrite if newer' doesn't think it needs to run?  'Overwrite if newer' compares the timestamp on the ebook file vs the reported date/time from the site.  But timezones can and will cause problems with that--timezones in Python are a headache that I'm not going to deal with.

Instead of 'overwrite if newer', use the 'update if new chapters' setting.  Not only will that compare chapter numbers instead of dates, it will reuse the chapters you've already downloaded for better performance.

**Comment by Twilight666** [comment]
Ok I got it. I didn't know about the date/time thing.

Also what I was doing is this:
1) I tried to update.. didn't work
2) Forced it to download
3) 5 minutes later the download is complete but I don't press yes yet
4) I try the update again andit is still not working.
5) I press yes and the new chapter is saved.

My point was that even those  5 minutes later, even though calibre could download the new chapter the update was still not working

Anyway I got my question/issue answered so thx!!
